activity_calendar/views.py

```diff
@@ -131,171 +38,292 @@ def check_join_constraints(request, parent_activity, recurrence_id):
+        return HttpResponseRedirect(self.request.get_full_path())
+
+    def get_failed_message(self, code):
+        if code == 'undefined' or code is None:
+            # Catch to prevent loops
+            return self.error_messages.get(code, 'Undefined error occured')
+
+        return self.error_messages.get(code, self.get_failed_message('undefined'))
+
+    def get_context_data(self, **kwargs):
+        kwargs = super(ActivityFormMixin, self).get_context_data(**kwargs)
+        kwargs.update({
+            'error_messages': self.error_messages,
+        })
+
+        return kwargs
+
+
+class ActivityMomentView(ActivityMixin, ActivityFormMixin, TemplateView):
+    pass
 
-            return redirect(request.get_full_path())
+
+class ActivitySimpleMomentView(LoginRequiredForPostMixin, FormMixin, ActivityMomentView):
+    form_class = RegisterForActivityForm
+    template_name = "activity_calendar/activity_page_no_slots.html"
+
+    error_messages = {
+        'undefined': _("Something went wrong. Please try again."),
+        'activity-full': _("This activity is already at maximum capacity. You can not subscribe to it."),
+        'invalid': _("You can not subscribe to this activity. Reason currently undefined"),
+        'already-registered': _("You are already registered for this activity"),
+        'not-registered': _("You are not registered to this activity"),
+        'closed': _("You can not subscribe, subscriptions are currently closed"),
+    }
+
+    def get_form(self, form_class=None):
+        # Intercept form creation for non-logged in users
+        if self.request.user.is_authenticated:
+            return super(ActivitySimpleMomentView, self).get_form(form_class=form_class)
+        else:
+            return None
+
+    def get_form_kwargs(self):
+        kwargs = {
+            # Add some set-up data based on the current situation
+            # This could be overwritten by the post data if supplied, which will yield the expected errors in that case
+            'data': {
+                'sign_up': not self.activity.is_user_subscribed(self.request.user, self.recurrence_id)
+            },
+            'activity': self.activity,
+            'recurrence_id': self.recurrence_id,
+            'user': self.request.user,
+        }
+        kwargs.update(super(ActivitySimpleMomentView, self).get_form_kwargs())
+        return kwargs
+
+    def get_context_data(self, **kwargs):
+        kwargs = super(ActivitySimpleMomentView, self).get_context_data(**kwargs)
+        kwargs.update({
+            'participants': self.activity.get_subscribed_participants(recurrence_id=self.recurrence_id),
+        })
+        return kwargs
+
+
+class ActivityMomentWithSlotsView(LoginRequiredForPostMixin, FormMixin, ActivityMomentView):
+    form_class = RegisterForActivitySlotForm
+    template_name = "activity_calendar/activity_page_slots.html"
+
+    error_messages = {
```

Comment:
`slot-full` should probably also be in this list; it currently displays "Undefined error occurred"